sources/api/prairiedog/src/bootconfig.rs

```diff
@@ -0,0 +1,410 @@
+    // Sequences of elements can mix quoted and unquoted values
+    // We also don't want to separate on a quoted comma
+    let mut elements = Vec::new();
+    let mut quote = None;
+    let mut expect_delimiter = false;
+    let mut start_index = 0;
+    for (i, c) in input.trim().chars().enumerate() {
+        if expect_delimiter && !c.is_whitespace() && c != ',' {
+            // return error::ExpectedArrayCommaSnafu { input: values }.fail();
+        }
+        if c == '\'' || c == '\"' {
+            if let Some(q) = quote {
+                // If the quote-types match, we're expecting a delimiter next
+                if q == c {
+                    quote = None;
+                    expect_delimiter = true;
+                }
+            } else {
+                quote = Some(c);
+            }
+        } else if c == ',' && quote.is_none() {
+            // We've encountered the delimiter, and if it's outside quotes, we have a new element
+            elements.push(parse_value(&input[start_index..i])?);
+            start_index = i + 1;
+            expect_delimiter = false;
+        }
+    }
+    ensure!(quote == None, error::UnbalancedQuotesSnafu { input });
+    // Push last element
+    let last_ele = if &input[start_index..] == "," {
+        // If it's just a comma, assume it's an empty value at the end
+        ""
+    } else {
+        &input[start_index..]
+    };
+    elements.push(parse_value(last_ele)?);
+    Ok(elements)
+}
+
+/// Takes a string representation of a bootconfig file and parse it into `BootSettings`
+fn parse_boot_config_to_boot_settings(bootconfig: &str) -> Result<BootSettings> {
+    let mut kernel_params: HashMap<BootConfigKey, Vec<BootConfigValue>> = HashMap::new();
+    let mut init_params: HashMap<BootConfigKey, Vec<BootConfigValue>> = HashMap::new();
+    for line in bootconfig.trim().lines() {
+        let mut kv = line.trim().split('=').map(|kv| kv.trim());
+        // Ensure the key is a valid boot config key
+        let key: BootConfigKey = kv
+            .next()
+            .ok_or(error::Error::InvalidBootConfig)?
+            .try_into()
+            .context(error::ParseBootConfigKeySnafu)?;
+        let values = parse_boot_config_values(kv.next().ok_or(error::Error::InvalidBootConfig)?)?;
+
+        if key != "kernel" && key.starts_with("kernel") {
+            kernel_params.insert(
+                key["kernel.".len()..]
+                    .try_into()
+                    .context(error::ParseBootConfigKeySnafu)?,
+                values,
+            );
+        } else if key != "init" && key.starts_with("init") {
+            init_params.insert(
+                key["init.".len()..]
+                    .try_into()
+                    .context(error::ParseBootConfigKeySnafu)?,
+                values,
+            );
+        } else if !key.starts_with("kernel") && !key.starts_with("init") {
+            return error::UnsupportedBootConfigKeySnafu { key }.fail();
+        }
```

Comment:
Should we always error in this case?
```suggestion
        } else {
            return error::UnsupportedBootConfigKeySnafu { key }.fail();
        }
```

Otherwise I think we'd allow input like:
```
kernel = foo
kernel.foo = bar
init = bar
init.bar = baz
```

Reply:
> kernel = foo
kernel.foo = bar
init = bar
init.bar = baz

I think we want to allow that. That's a valid boot config, it's just that later entries will overwrite the previous one.

I need that special case there for our default bootconfig which is just:
```
kernel = ""
init = ""
```

Reply:
Slight correction, it's not overwriting, but 
> sub-keys and a value can coexist under a parent key. For example, following config is allowed.:
foo = value1
foo.bar = value2

from https://www.kernel.org/doc/html/latest/admin-guide/bootconfig.html#same-key-values

Reply:
What I really should do here is special case allow the default bootconfig, but disallow `kernel` and `init` as parents keys for any non-null value, since `BootSettings` can't accommodate that.

So allow, `kernel = ""` but disallow `kernel = foo`